docs/controller_state_diagram.md

```diff
@@ -0,0 +1,123 @@
+# Controller state diagram
+
+This state diagram describes the state changes of Jobs from the
+perspective of the Controller.
+
+States in orange are pending (i.e. prior to a task being picked
+up by the Agent). States in red are final failed states, and those in green are final success states.
+
+Note that during the Agent-handled stages, task state is posted back to the controller in order to update Job state. The Agent does not itself update Job state.
+
+```mermaid
+stateDiagram
+    created: Created
+    initiated: Initiated
+    prepared: Prepared
+    executing: Executing
+    executed: Executed
+    finalized: Finalized
+    succeeded: Job Succeeded
+    error: Error
+    reset_reboot_dbmaintenance: Reset
+    retry: Retry after error
+    waiting: Waiting (Various)
+    cancelled: Cancelled
+    failed: Job Failed
+    dependency_failed: Dependency Failed
+    initiated_join: RUNJOB task created
+
+     [*] --> created: job created
+    created --> cancelled: RUNJOB task deactivated, CANCELJOB task created
+    created --> waiting
+    created --> dependency_failed
+    created --> reset_reboot_dbmaintenance
+
+    state initiated_join <<join>>
+    waiting --> initiated_join
+    created --> initiated_join
+    reset_reboot_dbmaintenance --> initiated_join
+    initiated_join --> initiated
+
+
+    initiated --> AgentStages
+    AgentStages: Agent
+    state AgentStages {
+        prepared --> executing
+
+    state execution_fail_fork_state <<choice>>
+    executing --> execution_fail_fork_state
+    execution_fail_fork_state --> executed: job runs to completion
+    execution_fail_fork_state --> error: job errors
+
+    error --> executed
+    executed --> finalized
+
+    }
+
+    state finalized_fork_state <<choice>>
+    AgentStages --> finalized_fork_state
+    finalized_fork_state --> succeeded
+    finalized_fork_state --> failed: fatal error
+    finalized_fork_state --> retry: non-fatal error
+
+    retry --> created
```

Comment:
wondering if it's worth clarifying how this works?
```suggestion
    retry --> created: RUNJOB task deactivated, controller restarted, new RUNJOB task created
```

Reply:
Is this how it works?  

For a controller error, if it's [fatal we mark the job as failed](https://github.com/opensafely-core/job-runner/blob/main/jobrunner/controller/main.py#L122), which doesn't deactivate the task but means the job won't be picked up again. If it's non-fatal, we don't do anything with the job or task, which I think should mean that the _same_ task gets picked up by the agent again.

For a [job error](https://github.com/opensafely-core/job-runner/blob/main/jobrunner/controller/main.py#L246), the task is already inactive because the error was a final agent state. If it's non-fatal we put it into a pending (retry) state so a new task gets created. But we don't restart the controller.

Reply:
yep you're right. we should split that into two states as you suggest, though.

Reply:
> I've made it go back to the initiated step

yep agree, I think this is best.